File: BEmu/IntradayTickRequest/MessageIntradayTick.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="IntradayTickRequest/MessageIntradayTick.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.IntradayTickRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class MessageIntradayTick : Message
    {
        private readonly ElementIntradayTickDataParent _parent;
        private readonly ElementIntradayTickResponseError _responseError;
        private readonly bool _isResponseError;

        internal MessageIntradayTick(CorrelationID corr, Service service, Dictionary<DateTime, Tuple<string, double, int>> ticks, bool includeConditionCodes) :
            base(new Name("IntradayTickResponse"), corr, service)
        {
            this._parent = new ElementIntradayTickDataParent(ticks, includeConditionCodes);
            this._responseError = null;
            this._isResponseError = false;
        }

        internal MessageIntradayTick(CorrelationID corr, Service service) :
            base(new Name("IntradayTickResponse"), corr, service)
        {
            this._parent = null;
            this._responseError = new ElementIntradayTickResponseError();
            this._isResponseError = true;
        }

        public override object this[string name, int index] { get { return null; } }
        public override string TopicName { get { return ""; } }
        public override int NumElements { get { return 1; } }
        public override Element AsElement { get { return new ElementReference(this); } }

        public override IEnumerable<Element> Elements
        {
            get
            {
                if (this._isResponseError)
                    yield return this._responseError;
                else
                    yield return this._parent;
            }
        }

        public override string ToString()
        {
            StringBuilder result = new StringBuilder();
            result.Append("IntradayTickResponse (choice) = {" + Environment.NewLine);

            foreach (var item in this.Elements)
            {
                result.Append(item.PrettyPrint(1));
            }

            result.Append("}");

            return result.ToString();
        }
    }
}

File: BEmuJava/src/BEmu/IntradayTickRequest/MessageIntradayTick.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.IntradayTickRequest.MessageIntradayTick.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package BEmu.IntradayTickRequest;

import BEmu.Element;
import BEmu.Message;
import BEmu.Service;
import BEmu.Name;
import BEmu.Datetime;
import BEmu.CorrelationID;

import java.util.Map;

public class MessageIntradayTick extends Message
{
	private final ElementIntradayTickDataParent _parent;
	
    MessageIntradayTick(CorrelationID corr, Map<Datetime, Tuple3<String, Double, Integer>> ticks, boolean includeConditionCodes, Service service)
    {
    	super(new Name("IntradayTickResponse"), corr, service);
        this._parent = new ElementIntradayTickDataParent(ticks, includeConditionCodes);
    }
    
	public Element getElement(String name) throws Exception
	{
		if(name.toLowerCase().equals(this._parent.name().toString().toLowerCase()))
			return this._parent;
		else
			throw new Exception("not implemented");
	}
	
	ElementIntradayTickDataParent firstElement()
	{
		return this._parent;
	}
    
	public String topicName()
	{
		return "";
	}
	
	public int numElements()
	{
		return 1;
	}
	
    public String toString()
    {
        StringBuilder result = new StringBuilder();
        result.append("IntradayTickResponse (choice) = {" + System.getProperty("line.separator"));
        result.append(this._parent.prettyPrint(1));
        result.append("}");

        return result.toString();
    }
}
